pkg/build/strategies/sti/sti.go

```diff
@@ -620,11 +646,295 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
+		return err
+	}
+
+	env := createBuildEnvironment(builder.config)
+	cmd := builder.createCommandForResultingImage(ctx.location)
+
+	ctx.labels = builder.createLabelsForResultingImage()
+
+	ctx.imageID, err = builder.commitContainer(ctx.containerID, cmd, user, env, ctx.labels)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (builder *STI) commitRuntimeImageStep(ctx *postExecutorStepContext) error {
+	glog.V(1).Info("Executing post execute step: commit runtime image")
+
+	user, err := builder.docker.GetImageUser(builder.config.RuntimeImage)
+	if err != nil {
+		return err
+	}
+
+	env := createBuildEnvironment(builder.config)
+
+	workDir, err := builder.docker.GetImageWorkdir(builder.config.RuntimeImage)
+	if err != nil {
+		return fmt.Errorf("Couldn't get working dir of %q image: %v", builder.config.RuntimeImage, err)
+	}
+
+	cmd := filepath.Join(workDir, api.Run)
+
+	ctx.labels = builder.createLabelsForResultingImage()
+
+	ctx.imageID, err = builder.commitContainer(ctx.containerID, cmd, user, env, ctx.labels)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (builder *STI) downloadAndExtractFile(artifactPath, artifactsDir, containerID string) error {
+	glog.V(2).Infof("Downloading file %q", artifactPath)
+
+	fd, err := ioutil.TempFile(artifactsDir, "s2i-runtime-artifact")
+	if err != nil {
+		return fmt.Errorf("Couldn't create temporary file for runtime artifact: %v", err)
+	}
+	defer func() {
+		fd.Close()
+		os.Remove(fd.Name())
+	}()
+
+	if err := builder.docker.DownloadFromContainer(artifactPath, fd, containerID); err != nil {
+		return fmt.Errorf("Couldn't download file (%q -> %q) from container %s: %v", artifactPath, fd.Name(), containerID, err)
+	}
+
+	// after writing to the file descriptor we need to rewind pointer to the beginning of the file before next reading
+	if _, err := fd.Seek(0, os.SEEK_SET); err != nil {
+		return fmt.Errorf("Couldn't seek to the beginning of the file %q: %v", fd.Name(), err)
+	}
+
+	if err := builder.tar.ExtractTarStream(artifactsDir, fd); err != nil {
+		return fmt.Errorf("Couldn't extract runtime artifact %q into the directory %q: %v", artifactPath, artifactsDir, err)
+	}
+
+	return nil
+}
+
+func (builder *STI) downloadFilesFromBuilderImageStep(ctx *postExecutorStepContext) error {
+	glog.V(1).Info("Executing post execute step: download files from the builder image")
+
+	artifactsDir := filepath.Join(builder.config.WorkingDir, api.RuntimeArtifactsDir)
+	if err := builder.fs.Mkdir(artifactsDir); err != nil {
+		return fmt.Errorf("Couldn't create directory %q: %v", artifactsDir, err)
+	}
+
+	for _, artifact := range builder.config.RuntimeArtifacts {
+		if err := builder.downloadAndExtractFile(artifact.Source, artifactsDir, ctx.containerID); err != nil {
+			return err
+		}
+
+		// for mapping like "/tmp/foo.txt -> app" we should create "app" and move "foo.txt" to that directory
+		dstSubDir := path.Clean(artifact.Destination)
+		if dstSubDir != "." && dstSubDir != "/" {
```

Comment:
@bparees:

>dstSubDir of "/" should be an error/warning. no absolute paths allowed since we aren't going to put the files into an absolute path

Now it's an error.

>> In the current implementation absolute path for destination is allowed but they're treated as the relative path. That happens because we always prepend the destination with a local path.
>
>what behavior do we do for the input from image flow? we should align with that.

I've copied behavior of input from image validation: https://github.com/openshift/origin/blob/d9dfa1005fe2684f722072bdd7e46a299d3c9a2d/pkg/build/api/validation/validation.go#L317-L325